pyiceberg/table/__init__.py

```diff
@@ -994,6 +1065,7 @@ def refs(self) -> Dict[str, SnapshotRef]:
         """Return the snapshot references in the table."""
         return self.metadata.refs
 
+    @table_commit_retry("properties")
```

Comment:
It depends on what we are trying to do here. There are two types of retries that we want to support:

- Intermittent network issues, catalog temporarily not available, etc.
- Retrying of commits because the table changed.

The first one should probably be done on the catalog level because we also need do differentiate between the different errors, and [see if they are retriable](https://github.com/apache/iceberg/blob/9c8e9ba67f41797ded61c11fa90af2d8df0da1cd/core/src/main/java/org/apache/iceberg/rest/ExponentialHttpRequestRetryStrategy.java#L91-L116).

For the second case, the one that you are solving here, we need some more logic around loading the latest version of the table. The retry is being done on the `CommitFailedException` which is thrown at a HTTP409 of the REST catalog. A 409 means a conflict and that the table has changed. At this point, we only support append and overwrite operations, which don't need any conflict detection. I believe that's Einstein's definition of insanity :)

Do we want to refresh the table metadata, and reapply the changes? I would expect like in https://github.com/apache/iceberg-python/blob/b1e33d48bc2c06a812a59c14e37ef73da9968f50/tests/catalog/test_sql.py#L803-L823 to fail after this change.

Reply:
My take on this is that if we try to implement retries in multiple places it will make it harder to prevent compounding retries, and thus result in a multiplying effect in the total number of retries, which is not a behaviour we want.

I would argue we should have just one place where we apply retries and it should keep the state of the number of attempts. 

We could catch multiple errors here instead of just `CommitFailedException` to account for network errors. 

In terms of accounting for table changes, that is a good point I had not originally considered. If the table has changed we would continue seeing the same error, in which case we need to perform a `refresh`. Refresh is an operation on the `Table` class, not the `Catalog`. So we need to have access to `Table`. With the current positioning of the retry decorator, we have access to the table instance so we can refresh the table.

What we can do is after every attempt we `refresh` the table. This is not optimal in the case where we failed due to network issues, since the table may have not changed. I would argue that the penalty that we are taking, in this case, would be minimal and is more than enough to pay for a more simplistic approach.